File: Builder/Builder01.js
// DEFINE AS PARTES DE UM CARRO

class Motor {
  constructor(tipo) {
    this.tipo = tipo;
  }
}

class Carroceria {
  constructor(estilo) {
    this.estilo = estilo;
  }
}

class Rodas {
  constructor(aro) {
    this.aro = aro;
  }
}

class Interior {
  constructor(cor) {
    this.cor = cor;
  }
}

// BUILDER

class CarroBuilder {
  constructor() {
    this.motor = null;
    this.carroceria = null;
    this.rodas = null;
    this.interior = null;
  }

  addMotor(tipo) {
    this.motor = new Motor(tipo);
    return this;
  }

  addCarroceria(estilo) {
    this.carroceria = new Carroceria(estilo);
    return this;
  }

  addRodas(aro) {
    this.rodas = new Rodas(aro);
    return this;
  }

  addInterior(cor) {
    this.interior = new Interior(cor);
    return this;
  }

  construir() {
    return new Carro(this.motor, this.carroceria, this.rodas, this.interior);
  }
}

// CONSTRUIR CARRO

class Carro {
  constructor(motor, carroceria, rodas, interior) {
    this.motor = motor;
    this.carroceria = carroceria;
    this.rodas = rodas;
    this.interior = interior;
  }

  mostrarDetalhes() {
    console.log(`
Especificações do carro:

Motor: ${this.motor.tipo}
Carroceria: ${this.carroceria.estilo}
Roda: ${this.rodas.aro}
Interior: ${this.interior.cor}
`);
  }
}

// CLIENTE

const carro1 = new CarroBuilder()
  .addMotor("1.4")
  .addCarroceria("Esportivo")
  .addRodas("18")
  .addInterior("Preto")
  .construir();

const carro2 = new CarroBuilder()
  .addMotor("1.0")
  .addCarroceria("Pickup")
  .addRodas("15")
  .addInterior("Rosa")
  .construir();

carro1.mostrarDetalhes();
carro2.mostrarDetalhes();

File: Builder/Builder01.py
# DEFINE AS PARTES DE UM CARRO

class Motor:
    def __init__(self, tipo):
        self.tipo = tipo


class Carroceria:
    def __init__(self, estilo):
        self.estilo = estilo


class Rodas:
    def __init__(self, aro):
        self.aro = aro


class Interior:
    def __init__(self, cor):
        self.cor = cor


# BUILDER

class CarroBuilder:
    def __init__(self):
        self.motor = None
        self.carroceria = None
        self.rodas = None
        self.interior = None

    def addMotor(self, tipo):
        self.motor = Motor(tipo)
        return self

    def addCarroceria(self, estilo):
        self.carroceria = Carroceria(estilo)
        return self

    def addRodas(self, aro):
        self.rodas = Rodas(aro)
        return self

    def addInterior(self, cor):
        self.interior = Interior(cor)
        return self

    def construir(self):
        return Carro(
            self.motor,
            self.carroceria,
            self.rodas,
            self.interior
        )


# CONSTRUIR CARRO

class Carro:
    def __init__(self, motor, carroceria, rodas, interior):
        self.motor = motor
        self.carroceria = carroceria
        self.rodas = rodas
        self.interior = interior

    def mostrar_detalhes(self):
        print(
            f"Especificações do carro: "
            f"Motor: {self.motor.tipo}, "
            f"Carroceria: {self.carroceria.estilo}, "
            f"Roda: {self.rodas.aro}, "
            f"Interior: {self.interior.cor}"
        )


# CLIENTE

carro1 = (
    CarroBuilder()
    .addMotor("1.4")
    .addCarroceria("Esportivo")
    .addRodas("18")
    .addInterior("Preto")
    .construir()
)

carro2 = (
    CarroBuilder()
    .addMotor("1.0")
    .addCarroceria("Pickup")
    .addRodas("15")
    .addInterior("Rosa")
    .construir()
)

carro1.mostrar_detalhes()
carro2.mostrar_detalhes()